torch/csrc/api/include/torch/data/datasets/map.h

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <torch/data/datasets/base.h>
+#include <torch/tensor.h>
+
+#include <ATen/core/ArrayRef.h>
+
+#include <cstddef>
+#include <utility>
+
+namespace torch {
+namespace data {
+namespace datasets {
+
+/// A `MapDataset` is a dataset that applies a transform to a source dataset.
+template <typename SourceDataset, typename AppliedTransform>
+struct MapDataset : BatchDataset<
+                        MapDataset<SourceDataset, AppliedTransform>,
+                        typename AppliedTransform::OutputBatchType> {
+  using DatasetType = SourceDataset;
+  using TransformType = AppliedTransform;
+  using OutputBatchType = typename TransformType::OutputBatchType;
+
+  MapDataset(DatasetType dataset, TransformType transform)
+      : dataset(std::move(dataset)), transform(std::move(transform)) {}
+
+  /// Gets a batch from the source dataset and applies the transform to it,
+  /// returning the result.
+  OutputBatchType get_batch(ArrayRef<size_t> indices) override {
+    return transform.apply_batch(dataset.get_batch(indices));
+  }
+
+  size_t size() const override {
+    return dataset.size();
+  }
+
+  SourceDataset dataset;
+  AppliedTransform transform;
+};
```

Comment:
It's a bit weird that mapping over a `Dataset` makes it a `BatchDataset` only (so you can no longer access individual elements).